[V1] Optimize handling of sampling metadata and req_ids list

njhill opened the pull request:
- Move the current `SamplingMetadata` object to a field in the persistent batch, updated only when the batch changes rather than constructed every step
- Keep `input_batch.req_ids` sized to the number of requests in the batch, so that anywhere that iterates over it doesn't need to slice (copy) the list or keep track of the separate request count. It is still updated in-place

njhill committed:
[V1] Optimize handling of sampling metadata and req_ids list

- Move SamplingMetadata to a field in the persistent batch, updated only when the batch changes rather than constructed every step
- Keep input_batch.req_ids sized to the number of requests in the batch, so that anywhere that iterates over it doesn't need to slice (copy) the list or keep track of the separate request count. It is still updated in-place

Signed-off-by: Nick Hill <nhill@redhat.com>

njhill commented on vllm/v1/worker/gpu_input_batch.py:
```diff
@@ -162,7 +163,6 @@ def __init__(
         self.stop_token_ids: List[Set[int]] = [
             set() for _ in range(max_num_reqs)
         ]
-        self.prompt_token_ids: Optional[torch.Tensor] = None
```
This was not actually used

WoosukKwon replied:
That's surprising 🤔

njhill commented:
@WoosukKwon this is the first step, I am working on follow-on simplification for the penalty parameters, etc.

njhill commented:
@WoosukKwon apologies, I am looking into the test failure.

njhill committed:
don't mutate "constant" sampling metadata tensors

Signed-off-by: Nick Hill <nhill@redhat.com>

njhill commented:
@WoosukKwon the test failure should be fixed now... the shared apply penalties code was doing in-place unsqueezes on the sampling penalty tensors - which I think is a bad thing to do but didn't cause a problem before because we were passing new slices every step.

njhill committed:
Merge remote-tracking branch 'origin/main' into sampler-streamline

# Conflicts:
#	vllm/v1/worker/gpu_input_batch.py

WoosukKwon commented:
Hi @njhill, do you mind if we merge #12193 first and review this PR? I'd like to prioritize the spec decode PR as it already got rebased many many times.

njhill commented:
@WoosukKwon that's fine with me.

njhill committed:
simplify sampling metadata

Signed-off-by: Nick Hill <nhill@redhat.com>

njhill committed:
Merge remote-tracking branch 'refs/remotes/origin/main' into sampler-streamline

Signed-off-by: Nick Hill <nhill@redhat.com>

# Conflicts:
#	tests/v1/worker/test_gpu_input_batch.py
#	vllm/v1/sample/sampler.py

njhill committed:
group stop_token_ids with min_tokens

Signed-off-by: Nick Hill <nhill@redhat.com>

njhill committed:
test updates

Signed-off-by: Nick Hill <nhill@redhat.com>

WoosukKwon commented:
@njhill Sorry for the delay. I will review this PR once it's rebased.

njhill committed:
Merge remote-tracking branch 'origin/main' into sampler-streamline

Signed-off-by: Nick Hill <nhill@redhat.com>

# Conflicts:
#	tests/v1/sample/test_sampler.py
#	tests/v1/worker/test_gpu_input_batch.py
#	vllm/v1/worker/gpu_input_batch.py
#	vllm/v1/worker/gpu_model_runner.py

njhill committed:
Merge remote-tracking branch 'origin/main' into sampler-streamline

Signed-off-by: Nick Hill <nhill@redhat.com>

njhill commented on vllm/v1/request.py:
```diff
@@ -46,7 +46,7 @@ def __init__(
         self._output_token_ids: List[int] = []
         self._all_token_ids: List[int] = self.prompt_token_ids.copy()
-        self.spec_token_ids: List[int] = []
+        self.spec_token_ids: Sequence[int] = []
```
These changes are so the type checker is happy with empty tuples as well as lists.

njhill committed:
Some more small list/tuple optimizations; fix linting

Signed-off-by: Nick Hill <nhill@redhat.com>

njhill commented on vllm/v1/core/scheduler.py:
```diff
@@ -195,8 +196,13 @@ def schedule(self) -> "SchedulerOutput":
                                              request.num_computed_tokens -
                                              request.num_tokens)
                 if num_scheduled_spec_tokens > 0:
+                    if isinstance(request.spec_token_ids, list):
+                        del request.spec_token_ids[num_scheduled_spec_tokens:]
```
In practice this branch will always be taken because it will be a list or empty tuple.

njhill commented:
@WoosukKwon I have now rebased. https://github.com/vllm-project/vllm/pull/13360 partially overlaps with this (e,g. I simplified some of the min_tokens handling in this one but have refactored completely in the other one based on the new abstraction). But I think it would be find to get this in first and I can rebase the other one if you're ok with that.

njhill committed:
Small adjustment

Signed-off-by: Nick Hill <nhill@redhat.com>

njhill commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -726,12 +724,11 @@ def _calc_spec_decode_metadata(
         scheduler_output: "SchedulerOutput",
         cu_num_tokens: np.ndarray,
-    ) -> Tuple[torch.Tensor, torch.Tensor]:
+    ) -> torch.Tensor:
```
This is just fixing incorrect return type.

WoosukKwon replied:
oh thanks for catching this!

njhill commented on vllm/v1/worker/gpu_input_batch.py:
```diff
@@ -394,87 +422,75 @@ def condense(self, empty_req_indices: List[int]) -> None:
+        del self._req_ids[self.num_reqs:]
+        del self.req_output_token_ids[self.num_reqs:]
+
+        # num_reqs entries should be non-None
+        assert all(req_id is not None
+                   for req_id in self._req_ids), "req_ids contains None"
```
This could be removed, not worth the cost of looping over all reqs just to check for a bug.

WoosukKwon replied:
I'm ok with either. Up to your preference :)

WoosukKwon commented:
@njhill I'm not sure it's worthwhile to change from `[]` to `()`.
I did a microbenchmark:
```python
N = 1024
x = []
# List
start = time.perf_counter()
for i in range(N):
    x.append([])
end = time.perf_counter()
print(f"list: {(end - start) * 1000:.3f} ms")

y = []
# Tuple
start = time.perf_counter()
for i in range(N):
    y.append(())
end = time.perf_counter()
print(f"tuple: {(end - start) * 1000:.3f} ms")
```
I find that adding 1024 (maximum number of requests in the batch) empty lists only takes 80-90 us. While using tuple reduces this time to 30-40 us, I think the 50 us gap (in the worst case) cannot justify the extra complexity here. When the batch size is 32, the gap becomes even smaller (7 us vs 2 us). WDYT?

njhill committed:
Fix rejection sampler test

Signed-off-by: Nick Hill <nhill@redhat.com>

njhill commented:
@WoosukKwon I agree it's not worth any extra complexity. Just might as well use `()` where it doesn't otherwise make any difference to the code. Let me check and revert where such changes were made..

WoosukKwon commented:
@njhill I think changing `List` to `Sequence` itself is increasing complexity? After that, we need to consider whether it's a tuple or list. I'd prefer to keep using `List` and `[]` if the performance is the only concern.

njhill committed:
Revert change related to list vs tuple

Signed-off-by: Nick Hill <nhill@redhat.com>

njhill commented:
@WoosukKwon sure, let me revert those too. I think mostly we don't need to consider the tuple/list difference because these are args or fields that would be considered read-only.

njhill committed:
Revert List->Sequence changes

Signed-off-by: Nick Hill <nhill@redhat.com>

njhill commented:
@WoosukKwon I need to fix up some of the gpu_model_runner tests, but I'll wait for your first review to make sure you are good with the changes overall before spending time on that.

WoosukKwon commented on vllm/v1/sample/metadata.py:
```diff
@@ -12,15 +12,13 @@ class SamplingMetadata:
     all_random: bool
-    rejection_sampling: bool
+
+    # The list will empty if no requests have spec tokens.
```
should we make it none in this case?

njhill replied:
Yes agree that's clearer.

WoosukKwon commented on vllm/v1/worker/gpu_input_batch.py:
```diff
@@ -394,87 +421,77 @@ def condense(self, empty_req_indices: List[int]) -> None:
+            logit_bias=self.logit_bias[:num_reqs],
         )
 
+    def get_sampling_metadata(
+            self,
+            req_id_to_spec_token_ids: Dict[str,
+                                           List[int]]) -> SamplingMetadata:
```
nit:
```suggestion
    def get_sampling_metadata(
        self,
        req_id_to_spec_token_ids: Dict[str, List[int]],
    ) -> SamplingMetadata:
```

WoosukKwon commented on vllm/v1/worker/gpu_input_batch.py:
```diff
@@ -333,13 +354,15 @@ def clear(self) -> None:
+        num_reqs = self.num_reqs
+        if num_reqs == 0:
             # The batched states are empty.
+            self.clear()
```
qq: Do we need this?

WoosukKwon reviewed (commented):
Amazing. Looks much cleaner! 😄

njhill commented on vllm/v1/core/scheduler.py:
```diff
@@ -195,8 +195,9 @@ def schedule(self) -> "SchedulerOutput":
                                              request.num_tokens)
                 if num_scheduled_spec_tokens > 0:
+                    del request.spec_token_ids[num_scheduled_spec_tokens:]
                     scheduled_spec_decode_tokens[request.request_id] = (
-                        request.spec_token_ids[:num_scheduled_spec_tokens])
+                        request.spec_token_ids)
```
It avoids creating a new list, just trims the existing one down to `num_scheduled_spec_tokens`, since any later spec token ids are essentially discarded anyhow.

WoosukKwon replied:
Got it! Maybe worth a comment.

njhill committed:
Address review comments

Signed-off-by: Nick Hill <nhill@redhat.com>

njhill committed:
Fix up gpu_model_runner tests

Signed-off-by: Nick Hill <nhill@redhat.com>

WoosukKwon reviewed (approved):
LGTM! Very nice simplification!

njhill committed:
Add comment

Signed-off-by: Nick Hill <nhill@redhat.com>
